### ISSUE TITLE
RouteValueDictionary prevents assembly from being unloaded

### ISSUE THREAD

**Issue by jasselin** [aspnet/AspNetCore#17509]
I am trying to load and unload an assembly containing controllers using the way explained here:
https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability

Having trouble unloading the assembly and having a look with WinDbg, I found that RouteValueDictionary, probably used by the endpoint middleware is holding references to types in the assembly, which seems to prevent it from being unloaded.

`0000027274F313E0 (pinned handle)
    -> 00000275A5223050 System.Object[]
    -> 000002747526DD00 System.Collections.Concurrent.ConcurrentDictionary``2[[System.Type, System.Private.CoreLib],[Microsoft.Extensions.Internal.PropertyHelper[], Microsoft.AspNetCore.Mvc.Core]]
    -> 000002747526E018 System.Collections.Concurrent.ConcurrentDictionary``2+Tables[[System.Type, System.Private.CoreLib],[Microsoft.Extensions.Internal.PropertyHelper[], Microsoft.AspNetCore.Mvc.Core]]
    -> 000002747526DF08 System.Collections.Concurrent.ConcurrentDictionary``2+Node[[System.Type, System.Private.CoreLib],[Microsoft.Extensions.Internal.PropertyHelper[], Microsoft.AspNetCore.Mvc.Core]][]
    -> 00000274752709B0 System.Collections.Concurrent.ConcurrentDictionary``2+Node[[System.Type, System.Private.CoreLib],[Microsoft.Extensions.Internal.PropertyHelper[], Microsoft.AspNetCore.Mvc.Core]]
    -> 000002747526B968 System.RuntimeType
    -> 000002747526B2C0 System.Reflection.LoaderAllocator`

Is there a way to remove this reference or am I stuck with the assembly in memory?

**Comment by javiercn** [comment]
@jasselin thanks for contacting us.

It is likely we don't support what you mention. While you can unload assemblies in .NET Core, loading and unloading controllers, etc is not something we have explicit support as far as I know.

@rynowak might have some suggestions here, but in principle I would say this is not something we support.

**Comment by davidfowl** [comment]
> It is likely we don't support what you mention. While you can unload assemblies in .NET Core, loading and unloading controllers, etc is not something we have explicit support as far as I know.

I don't think this is true. While we didn't explicitly do any work to support this scenario we should be aware of the things we do that would prevent it from working. This may just be the tip of iceberg so we should keep an eye out as more people build more plugin like systems like this.

I could imagine a universe where we do some work here to avoid this cache if we're not in the default load context and avoid this code path for internal usage (if we don't already).

**Comment by javiercn** [comment]
We've moved this issue to the Backlog milestone. This means that it is not going to happen for the coming release. We will reassess the backlog following the current release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by davidfowl** [comment]
Isn't this labeled incorrectly? Shouldn't it be in servers?

**Comment by jasselin** [comment]
@davidfowl Thanks for considering the change, it's a bit sad that I am stuck with this for 3.1, but I get that it is pretty late to introduce this kind of change.

BTW, I am really looking forward to see what you bring into the product to support hot reloading. I am doing a hacky workaround compiling dynamic assemblies and the only missing puzzle piece is unloading those assemblies. I guess I'll have to tolerate a memory leak for now! 🙃
